documentation-support/src/main/java/tech/picnic/errorprone/documentation/BugPatternTestExtractor.java

```diff
@@ -0,0 +1,179 @@
+
+    private final List<String> encounteredClasses = new ArrayList<>();
+
+    boolean hasTestUsingClassInstance(String clazz) {
+      return encounteredClasses.contains(clazz);
+    }
+
+    @Override
+    public @Nullable Void visitMethodInvocation(MethodInvocationTree node, VisitorState state) {
+      if (BUG_PATTERN_TEST_METHOD.matches(node, state)) {
+        MemberSelectTree firstArgumentTree = (MemberSelectTree) node.getArguments().get(0);
+        encounteredClasses.add(firstArgumentTree.getExpression().toString());
+      }
+      return super.visitMethodInvocation(node, state);
+    }
+  }
+
+  private static final class CollectBugPatternTests
+      extends TreeScanner<@Nullable Void, @Nullable Void> {
+    private static final Matcher<ExpressionTree> IDENTIFICATION_SOURCE_LINES =
+        instanceMethod()
+            .onDescendantOf("com.google.errorprone.CompilationTestHelper")
+            .named("addSourceLines");
+    private static final Matcher<ExpressionTree> REPLACEMENT_INPUT =
+        instanceMethod()
+            .onDescendantOf("com.google.errorprone.BugCheckerRefactoringTestHelper")
+            .named("addInputLines");
+    private static final Matcher<ExpressionTree> REPLACEMENT_OUTPUT =
+        instanceMethod()
+            .onDescendantOf("com.google.errorprone.BugCheckerRefactoringTestHelper.ExpectOutput")
+            .named("addOutputLines");
+
+    private final VisitorState state;
+    private final List<String> identificationTests = new ArrayList<>();
+    private final List<BugPatternReplacementTestDocumentation> replacementTests = new ArrayList<>();
+
+    @Var private String replacementOutputLines = "";
+
+    CollectBugPatternTests(VisitorState state) {
+      this.state = state;
+    }
+
+    public ImmutableList<String> getIdentificationTests() {
+      return ImmutableList.copyOf(identificationTests);
+    }
+
+    public ImmutableList<BugPatternReplacementTestDocumentation> getReplacementTests() {
+      return ImmutableList.copyOf(replacementTests);
+    }
+
+    @Override
+    public @Nullable Void visitMethodInvocation(MethodInvocationTree node, @Nullable Void unused) {
+      if (IDENTIFICATION_SOURCE_LINES.matches(node, state)) {
+        identificationTests.add(getSourceLines(node));
+      } else if (REPLACEMENT_INPUT.matches(node, state)) {
+        /* The visitor starts with `addOutputLines` and in the next visit it will go over the `addInputLines`. */
+        replacementTests.add(
+            BugPatternReplacementTestDocumentation.create(
+                getSourceLines(node), replacementOutputLines));
+      } else if (REPLACEMENT_OUTPUT.matches(node, state)) {
+        replacementOutputLines = getSourceLines(node);
+      }
+      return super.visitMethodInvocation(node, unused);
+    }
+
+    // XXX: Duplicate from `ErrorProneTestSourceFormat`, should we move this to `SourceCode` util?
+    private static String getSourceLines(MethodInvocationTree tree) {
```

Comment:
> should we move this to `SourceCode` util? 

Didn't apply something like this yet, to not introduce a dependency on `error-prone-contrib`. I think we should have a more generic "utilities" module.